text/0225-configuration.md

```diff
@@ -0,0 +1,329 @@
+
+SDKs validating configuration *MUST* fail when they encounter a configuration with an unsupported version. Generally, this means fail when encountering a major version which is not recognized. An SDK might choose to maintain a library of validators / parsers for each major version, and use the configuration version to select and use the correct instance. Differences in minor versions (except pre-1.0 minor versions) SHOULD be acceptable, with the caveat that allowable property additions and removals MAY result in configuration that is different than excepted.
+
+## Trade-offs and mitigations
+
+### Additional method to configure OpenTelemetry
+
+If the implementation suggested in this OTEP goes ahead, users will be presented with another mechanism for configuring OpenTelemetry. This may cause confusion for users who are new to the project. It may be possible to mitigate the confusion by providing users with best practices and documentation.
+
+### Many ways to configure may result in users not knowing what is configured
+
+As there are multiple mechanisms for configuration, it's possible that the active configuration isn't what was expected. This could happen today, and one way it could be mitigated would be by providing a mechanism to list the active OpenTelemetry configuration.
+
+### Errors or difficulty in configuration files
+
+Configuration files provide an opportunity for misconfiguration. A way to mitigate this would be to provide clear messaging and fail quickly when misconfiguration occurs.
+
+## Prior art and alternatives
+
+The working group looked to the OpenTelemetry Collector and OpenTelemetry Operator for inspiration and guidance.
+
+### Alternative schema languages
+
+In choosing to recommend JSON schema, the working group looked at the following options:
+
+* [Cue](https://cuelang.org/) - A promising simpler language to define a schema, the working group decided against CUE because:
+  * Tooling available for validating CUE files in languages outside of Go were limited.
+  * Familiarity and learning curve would create problems for both users and contributors of OpenTelemetry.
+* [Protobuf](https://developers.google.com/protocol-buffers) - With protobuf already used heavily in OpenTelemetry, the format was worth investigating as an option to define the schema. The working group decided against Protobuf because:
+  * Validation errors are the result of serlization errors which can be difficult to interpret.
+  * Limitations in the schema definition language result in poor ergonomics if type safety is to be retained.
+
+## Open questions
+
+### How to handle environment variable / file config overlap?
+
+How does file configuration interact with environment variable configuration when both are present?
+
+* Solution 1: Ignore environment configuration when file configuration is present. Log a warning to the user indicating that multiple configuration modes were detected, but use the file configuration as the source of truth.
```

Comment:
I think we could log a warning when we detect this, but failing is too strong. Consider the implications if a user is operating in an environment they don't fully control (i.e. where an ops team configures environment variables by default which they extend / layer on top of).

Reply:
I think suggestion #1 as mentioned in this [comment](https://github.com/open-telemetry/oteps/pull/225#discussion_r1116269308) in combination with env var substitution supports most cases.

It may be hard to differentiate when it's appropriate to log warnings. Consider the following config:

```yaml
otlp:
    endpoint: ${env:OTEL_EXPORTER_OTLP_TRACES_ENDPOINT}
```

Using the config file and setting `OTEL_EXPORTER_OTLP_TRACES_ENDPOINT` would in theory result in a case that should warn, but the user just happens to be using the same env variable in their configuration. Warning here would require handling the case where some env vars listed in the specification must be known to the configuration parsing layer, or the code handling env variables must be aware of the configuration parser somehow.

Reply:
Somethings can be left to the user to resolve :). If they don't want a warning printed in the case they are doing something like this it can be on them to use a different variable name.

Reply:
Tho I'm of the opinion there should be no warnings and environment variables should simply override anything in the config file.

Reply:
> Consider the implications if a user is operating in an environment they don't fully control (i.e. where an ops team configures environment variables by default which they extend / layer on top of).

That's a fair point, although if what we want is for people to migrate to this new configuration format I don't see this as a bad thing necessarily (I guess it depends on how costly we deem the migration to be)

Reply:
Captured the proposed solutions as a decision that must be made in the specification https://github.com/open-telemetry/oteps/pull/225/commits/19444556d823cfedd2643337f7f12633e959830e